shards/egui/src/visual/mod.rs

```diff
@@ -0,0 +1,2521 @@
+                  custom_state: None,
+                }
+              });
+
+              x.params.as_mut().unwrap().push(param_to_add);
+            }
+
+            // set flag to true
+            x.get_custom_state::<FunctionState>().unwrap().params_sorted = true;
+          }
+
+          egui::ScrollArea::both().show(self.ui, |ui| {
+            for (idx, param) in params.into_iter().enumerate() {
+              let param = param.as_table().unwrap();
+              let name: &str = param
+                .get_fast_static("name")
+                .try_into()
+                .expect("A shard's parameter name must be a string!");
+              let help_text: &str = param
+                .get_fast_static("help")
+                .try_into()
+                .expect("A shard's parameter help text must be a string!");
+              let types = param
+                .get_fast_static("types")
+                .as_seq()
+                .expect("A shard's parameter types must be a sequence!");
+              let help_text = if help_text.is_empty() {
+                "No help text provided."
+              } else {
+                help_text
+              };
+              egui::CollapsingHeader::new(name)
+                .default_open(true)
+                .show(ui, |ui| {
+                  ui.horizontal(|ui| {
+                    // button to reset to default
+                    if ui
+                      .button(emoji("🔄"))
+                      .on_hover_text("Reset to default value.")
+                      .clicked()
+                    {
+                      // reset to default
+                      shlog_debug!("Resetting: {} to default value.", name);
+                      let default_value = param.get_fast_static("default");
+                      x.params.as_mut().map(|params| {
+                        params[idx].value = var_to_value(&default_value).unwrap();
+                      });
+                      self.context.has_changed = true;
+                    }
+                    if ui
+                      .button(emoji("🔧"))
+                      .on_hover_text("Change value type.")
+                      .clicked()
+                    {
+                      // let (sender, receiver) = mpsc::channel();
+                      // let query = Var::ephemeral_string("").into();
+                      // get_global_visual_shs_channel_sender()
+                      //   .send((query, sender))
+                      //   .unwrap();
+                      // x.get_custom_state::<FunctionState>().unwrap().receiver =
+                      //   Some(UniqueReceiver::new(receiver));
+
+                      // open a dialog to change the value
+
+                      let mouse_pos = ui
+                        .ctx()
+                        .input(|i| i.pointer.hover_pos().unwrap_or_default());
+                      self.context.swap_state = Some(SwapState::Param(ParamSwapState {
```

Comment:
Is this for the context menu? I think egui already has that with open state tracking and all.
You can just have a button and then:
```rs
let response = ui.button( .... );
response.context_menu(|ui| {
  ...
});
```

Reply:
Not sure what you mean because the end of the operation will happen on different frames anyways.
This is literally when we are selecting a new shard  or replace